src/scanpy/get/_aggregated.py

```diff
@@ -554,8 +554,9 @@ def sparse_indicator(
         weight = mask * weight
     elif mask is None and weight is None:
         weight = np.broadcast_to(1.0, len(categorical))
+    missing = categorical.isna()
     a = sparse.coo_matrix(
-        (weight, (categorical.codes, np.arange(len(categorical)))),
+        (weight[~missing], (categorical.codes[~missing], np.arange((~missing).sum()))),
         shape=(len(categorical.categories), len(categorical)),
```

Comment:
```suggestion
        shape=(len(categorical.categories), sum(~missing)),
```
or something like this? I would think the shape should change too if the shape of the arguments changes

Reply:
Good point, but, then we’d also have to subset `self.data`. And then that raises the question why we don’t already do that using `mask`: is it faster to mask than to subset?

If we want to not slice, we could do this before creating the matrix:

```py
if (missing := groupby.isna()).any():
    mask = mask & ~missing if mask is not None else ~missing
```

and then pass `np.where(mask, categorical.codes, 0)` to `coo_matrix` (to avoid `-1`s tripping it up)

but I wonder why we don’t just apply the mask to the data …